README.md
<p align='right'>
    <a href="https://badge.fury.io/js/veams-utility-type-ahead"><img src="https://badge.fury.io/js/veams-utility-type-ahead.svg" alt="npm version" height="18"></a>
    <a href='https://gitter.im/Sebastian-Fitzner/Veams?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge'><img src='https://badges.gitter.im/Sebastian-Fitzner/Veams.svg' alt='Gitter Chat' /></a>
</p>

# TypeAhead

This utility offers a type-ahead function that can be applied to input elements.

It generates a list with keywords fetched from a server.

-----------

## Requirements
- `Veams >= v5.0.0` - Veams Framework.

-----------

## Installation

### Installation with Veams

`veams install vu type-ahead`

-------------

## Fields

### type-ahead.hbs

#### Settings

- settings.typeAheadClasses {`String`} - _Modifier classes._
- settings.typeAheadJsOptions {`Object`} - _Options object which gets stringified._
- settings.height {`Number`} - _Height of suggestion list._

-------------

## JavaScript Options

The module gives you the possibility to override default options:

- currentValueClass {`String`} ['type-ahead__current-value'] - _Current value class._
- inputSel {`String`} [[data-js-item="type-ahead__input"]] - _Element selector for input._
- minLength {`Number`} [2] - _Minimal length of input value before suggestions will be shown._
- minLoadTime {`Number`} [0] - _Minimal time that must have passed before new request is sent to server._
- suggestionItemSel {`String`} [[data-js-item="suggestion-item"]] - _Element selector for suggestion item._
- suggestionItemSel {`String`} [[data-js-item="type-ahead__list"]] - _Element selector for type-ahead list._
- templates.tplSuggestionsA11y {`String`} ['SUGGESTIONS__A11Y'] - _Template name for suggestion (list item)._
- templates.tplSuggestionsOpt {`String`} ['SUGGESTIONS__OPTIONS'] - _Template name for option._

CHANGELOG.md
### v5.1.0
- Make component compatible with veams-cli@2.0.xx

### v5.0.0
- Update folder structure to be in sync with latest generator output
- Remove import snippets from INSERTPOINTS.md

### v4.0.0
- Update insertpoints paths for new folder structure (used in generator-veams v9.x)
- Unify components folder structure

### v3.1.4
- Fix version conflict

### v3.1.3
- Fix click event issue if veams-query is used

### v3.1.2
- Update version in JS

### v3.1.1
- Fix insertpoints
- Rename `typeahead` to `type-ahead`
- Update readme

### v3.1.0
- Update source code to support Veams 5 framework
- Rewrite module

### v3.0.1
- use `null` instead of `false` for empty values in options

### v3.0.0
- make module compatible with VeamsQuery

### v2.0.0
- rewrite in ES6
- append type ahead list to parent form or element itself

### v1.0.0
- initial version

js/typeahead.js
/**
 * This module offers type-ahead functionality for input fields.
 *
 * @module Typeahead
 * @version v3.1.0
 *
 * @author José Medina
 */

// Imports
import {Veams} from 'app';

import VeamsComponent from 'veams/lib/common/component';
import AjaxService from 'veams/lib/services/http';
import Toggler from '../../../components/toggler/js/toggler';

// Variables
const $ = Veams.$;
const Helpers = Veams.helpers;

class Typeahead extends VeamsComponent {

	/**
	 * Constructor for our class
	 *
	 * @see module.js
	 *
	 * @param {Object} obj - Object which is passed to our class
	 * @param {Object} obj.el - element which will be saved in this.el
	 * @param {Object} obj.options - options which will be passed in as JSON object
	 */
	constructor(obj) {
		let options = {
			currentValueClass: "typeahead__current-value",
			inputSel: '[data-js-item="typeahead__input"]',
			minLength: 2,
			minLoadTime: 0,
			suggestionItemSel: '[data-js-item="suggestion-item"]',
			suggestionListSel: '[data-js-item="typeahead__list"]',
			templates: {
				tplSuggestionsA11y: "SUGGESTIONS__A11Y",
				tplSuggestionsOpt: "SUGGESTIONS__OPTIONS"
			}
		};

		super(obj, options);
	}


	/**
	 * Get module information
	 */
	static get info() {
		return {
			version: '3.1.0'
		};
	}


	/**
	 * Subscribe handling
	 */
	get subscribe() {
		return {
		}
	}


	/**
	 * Initialize the view
	 *
	 */
	initialize() {
		this.$suggestionList = this.$el.find(this.options.suggestionListSel);
		this.$input = this.$el.find(this.options.inputSel);
		this.currentValue = this.$input.val();

		this.suggestionList = new Toggler({
			el: this.$suggestionList[0],
			namespace: 'Toggler',
			appInstance: Veams
		});

		this.suggestionList.close();

		this.service = new AjaxService(this.options.serviceOptions);

		this.service.requestDidOpen = (request, obj) => {
			if (this.options.serviceOptions.contentType) {
				request.setRequestHeader("Content-type", this.options.serviceOptions.contentType);
			}
			request.setRequestHeader("X-CSRFToken", this.options.serviceOptions["X-CSRFToken"]);
		}
	}


	/**
	 * Bind Vent events for the resize events and the load more button
	 */
	bindEvents() {
		this.$input.on(Veams.EVENTS.keyup, this.onKeyup.bind(this));

		$(window).on(Veams.EVENTS.keydown, this.onWindowKeydown.bind(this));

		this.$input.on(Veams.EVENTS.blur, this.onItemBlur.bind(this));

		this.$suggestionList
			.on(Veams.EVENTS.touchstart, this.options.suggestionItemSel, this.onChosenSuggestion.bind(this))
			.on(Veams.EVENTS.click, this.options.suggestionItemSel, this.onChosenSuggestion.bind(this));
	}


	/**
	 * Set focus when needed
	 *
	 * @param {Object} obj - CTA event object
	 */
	setFocus(obj) {

		if (obj.isActive) {

			setTimeout(() => {
				this.$input[0].focus();
			}, 1000);
		}
	}


	/**
	 * Handle window key down event to control the module with arrow keys, enter and escape
	 *
	 * @param {Object} event - Event object
	 */
	onWindowKeydown(event) {
		if (this.el.contains(event.target)) {
			let getFocusableElements = () => {
				return Array.prototype.slice.call($("a, input", this.$el));
			};
			let focusableEls, index;

			switch (event.keyCode) {
				case 38:
					event.preventDefault();
					focusableEls = getFocusableElements();
					index = focusableEls.indexOf(document.activeElement);

					if (index > -1 && index - 1 >= 0) {
						focusableEls[index - 1].focus();
					}
					break;
				case 40:
					event.preventDefault();
					focusableEls = getFocusableElements();
					index = focusableEls.indexOf(document.activeElement);

					if (index > -1 && index + 1 < focusableEls.length) {
						focusableEls[index + 1].focus();
					}
					break;
				case 27:
					this.closeSuggestionList();
					break;
				case 13:
					this.triggerChosenSuggestion(event.target);
					this.closeSuggestionList();
					break;
			}
		}
	}


	/**
	 * Handle input key up event to control the module with arrow keys, enter and escape
	 *
	 * @param {Object} event - Event object
	 */
	onKeyup(event) {
		switch (event.keyCode) {
			case 38:
			case 40:
			case 27:
			case 13:
				event.preventDefault();
				break;
			default:
				this.currentValue = this.$input.val();
				this.render(this.generateTimeOfRequest());
		}
	}


	/**
	 * Bind filter items events as soon as render occurs
	 *
	 */
	bindItemEvents() {
		this.$suggestionList.find(this.options.suggestionItemSel)
			.on(Veams.EVENTS.blur, this.onItemBlur.bind(this))
			.on(Veams.EVENTS.focus, this.onItemFocus.bind(this))
			.on(Veams.EVENTS.mouseenter, this.onItemMouseEnter.bind(this))
			.on(Veams.EVENTS.mouseleave, this.onItemMouseLeave.bind(this));
	}


	/**
	 * Handle touch and click event
	 *
	 * @param {Object} event - Event object
	 */
	onChosenSuggestion(event) {
		let target = event.currentTarget;
		event.preventDefault();

		this.triggerChosenSuggestion(target);

		this.closeSuggestionList();
	}


	/**
	 * Handle item focus event
	 *
	 * @param {Object} event - Event object
	 */
	onItemFocus(event) {
		this.$input.val($(event.target).closest(this.options.suggestionItemSel).attr('data-value'));
	}


	/**
	 * Handle item mouse enter event
	 *
	 * @param {Object} event - Event object
	 */
	onItemMouseEnter(event) {
		this.onItemFocus.apply(this, arguments);
	}


	/**
	 * Handle item mouse leave event
	 */
	onItemMouseLeave() {
		this.$input.val(this.currentValue);
	}


	/**
	 * Handle item blur event
	 *
	 * @param {Object} event - Event object
	 */
	onItemBlur(event) {
		this.$input.val(this.currentValue);

		if (!!event.relatedTarget && !this.$suggestionList[0].contains(event.relatedTarget) &&
			event.relatedTarget !== this.$input[0]) {
			this.closeSuggestionList();
		}
	}


	/**
	 * Helper function that looks for the current input value and transmit it
	 *
	 * @param {Object} target - DOM Node
	 */
	triggerChosenSuggestion(target) {
		this.currentValue = $(target).data("value") || $(target).val();
		this.$input.val(this.currentValue);

		this.$el.trigger(Veams.EVENTS.typeahead.suggestionChoosen, {
			search: this.currentValue
		});
	}


	/**
	 * Render class
	 *
	 * Perform and server request to get the suggestions to a given value
	 *
	 * @param {Number} timeOfRequest - Time of request
	 */
	render(timeOfRequest) {
		if (this.currentValue.length < this.options.minLength || !this.$el[0].contains(document.activeElement)) {
			this.closeSuggestionList();
			return false;
		}

		return new Promise((resolve, reject) => {
			this.service[this.options.serviceOptions.method]({
				data: JSON.stringify({}),
				url: this.options.serviceOptions.url + "?typeahead=" + this.currentValue
			}).then((data) => {
				setTimeout(() => {
					if (timeOfRequest !== this.timeOfRequest) {
						return false;
					}

					let $suggestionsItems = $(
						Veams.templater.render(this.options.templates.tplSuggestionsA11y, this.formatResponseDataHelper(data)));
					this.$suggestionList.html($suggestionsItems);

					this.bindItemEvents();

					this.suggestionList.open();

					resolve();
				}, Math.max(this.options.minLoadTime - (Date.now() - timeOfRequest), 0));
			}, (error) => {
				reject();
				throw new Error("Typeahead render() :: could not resolve server request :: ", error);
			});
		});
	}


	/**
	 * Cache the current timestamp to figure out which was the last request
	 *
	 * If 2 delta time between request is less than minLoadTime, they will be rejected
	 * and the one with the latest timestamp will be sent
	 *
	 * @return {Number} - Time of request
	 */
	generateTimeOfRequest() {
		this.timeOfRequest = Date.now();

		return this.timeOfRequest;
	}


	/**
	 * Data formatter that prepares the data for the template to render
	 *
	 * @param {Object} data - Data to be prepared for rendering
	 * @return {Object} - Prepared data (ready for rendering)
	 */
	formatResponseDataHelper(data) {
		data.suggestions = data.suggestions.map((item) => {
			item.suggestionText = this.getSuggestionTextHelper(item.name);
			item.searchVal = this.currentValue;

			return item;
		});

		return data;
	}


	/**
	 * Data formatter that prepares the data for the template to render
	 *
	 * @param {String} name - Item name
	 * @return {String} - Modified data
	 */
	getSuggestionTextHelper(name) {
		let currentValueCaseInsensitive = new RegExp(this.currentValue, "ig");

		return name.replace(currentValueCaseInsensitive, '<em class="' + this.options.currentValueClass + '">$&</em>');
	}


	/**
	 * Close the suggestion list and empty the list
	 *
	 */
	closeSuggestionList() {
		this.suggestionList.close();
		this.$suggestionList.empty();
	}
}

export default Typeahead;


js/views/type-ahead-view.js
/**
 * Represents a TypeAheadView.
 * @class TypeAheadView
 *
 * @module TypeAheadView
 * @version v3.0.1
 *
 * Use the get keyword to make our methods serve as getters for a property.
 * This means they will be accessible as properties, but defined as methods,
 * retaining compatibility with any existing references if you're converting existing code.
 *
 * @author Andy Gutsche
 */

import App from '../../../app';
import Helpers from '../../../utils/helpers';
import Tpl from '../../../templates/templates';
import TypeAheadCollection from '../collections/type-ahead-collection';
import TypeAheadItemView from './type-ahead-item-view';

let Handlebars = require('handlebars/runtime')['default'];
let $ = App.$;
let Template = Tpl(Handlebars);

// Creates a new view class object
class TypeAheadView extends App.ComponentView {

	/**
	 * Get template
	 *
	 */
	get template() {
		return this._template;
	}

	/**
	 * Set template
	 *
	 */
	set template(tpl) {
		this._template = tpl;
	}

	/**
	 * Get options
	 *
	 */
	get _options() {
		return {
			appendTarget: null,
			contextClass: null,
			deleteBtn: '[data-js-atom="delete-btn"]',
			form: '[data-js-atom="form"]',
			url: null,
			inputField: '[data-js-atom="input-field"]',
			item: '[data-js-atom="type-ahead-item"]',
			itemClass: 'type-ahead__item',
			list: '[data-js-atom="type-ahead-list"]',
			modifierClass: null,
			template: Template['TYPEAHEAD'],
			threshold: 4
		}
	}

	/**
	 * Set options
	 *
	 */
	set _options(opts) {
		this.options = opts;
	}

	/**
	 * Get module information
	 */
	static get info() {
		return {
			name: 'TypeAhead',
			version: '3.0.1',
			vc: true
		};
	}

	/**
	 * Initialize the view and merge options
	 *
	 * @param {Object} obj - options obj
	 * @public
	 */
	initialize(obj) {
		this.$el = $(this.el); //TODO: set via AppComponentView
		this._options = Helpers.defaults(obj.options || {}, this._options);
		this.template = this.options.template;

		App.registerModule && App.registerModule(TypeAheadView.info, this.el);

		this.$inputField = $(this.options.inputField, this.$el);
		this.$form = this.$el.closest(this.options.form);
		this.$appendTarget = this.options.appendTarget ? $(this.options.appendTarget, this.$el) : this.$el;

		this.bindEvents();
	}

	/**
	 * Bind all events
	 *
	 * @public
	 */
	bindEvents() {
		let fnResetInput = this.resetInput.bind(this);
		let fnFetchData = this.fetchData.bind(this);
		let fnRemoveTypeAhead = this.removeTypeAhead.bind(this);
		let fnSearch = this.search.bind(this);
		let fnCalculateWidthAndPos = this.calculateWidthAndPos.bind(this);

		// global events
		App.Vent.on(App.EVENTS.typeAhead.search, fnSearch);
		App.Vent.on(App.EVENTS.resize, fnCalculateWidthAndPos);

		// local events
		this.$el.on(App.EVENTS.click, this.options.deleteBtn, fnResetInput);
		this.$el.on(App.EVENTS.keyup, this.options.inputField, fnFetchData);
		this.$el.on(App.EVENTS.blur, this.options.inputField, fnRemoveTypeAhead);
	}

	/**
	 * Fetch auto suggest data
	 *
	 * @private
	 */
	fetchData() {

		if (this.$inputField.val().length >= this.options.threshold) {
			this.collection = new TypeAheadCollection();

			this.collection.fetch({
				url: this.options.url,
				success: () => {
					this.render();
				}
			});
		} else {
			this.removeTypeAhead(true);
		}
	}

	/**
	 * Reset input
	 *
	 * @private
	 */
	resetInput() {
		this.$inputField.val('');
		this.removeTypeAhead(true);
	}

	/**
	 * Submit form
	 *
	 * @param {Object} e - event object
	 * @private
	 */
	search(e) {
		this.$inputField.val(e.keyword);
		this.$form[0].submit();
	}

	/**
	 * Remove type ahead element
	 *
	 * @param {Object} e - event object
	 * @private
	 */
	removeTypeAhead(e) {
		let timeout = typeof e === 'boolean' && e ? 0 : 200;

		clearTimeout(this.timeout);

		this.timeout = setTimeout(() => {

			if (this.$typeAheadEl && this.$typeAheadEl.length) {
				this.$typeAheadEl.remove();
				this.$typeAheadEl = null;
			}
		}, timeout);
	}

	/**
	 * Calculate width and position of type ahead element
	 *
	 * @param {Object} e - event object
	 * @private
	 */
	calculateWidthAndPos(e) {
		let inputOffset = this.$inputField.offset();

		if (!this.$typeAheadEl || !this.$typeAheadEl.length) {
			return;
		}

		if (e && !App.support.touch) {
			this.removeTypeAhead(true);
		}

		this.$typeAheadEl.css({
			width: this.$inputField.outerWidth(),
			top: inputOffset.top + this.$inputField.outerHeight(),
			left: inputOffset.left
		});
	}

	/**
	 * Render one item
	 *
	 * @param {Object} model - item model
	 * @private
	 */
	renderOne(model) {
		let item = new TypeAheadItemView({
			model: model,
			tagName: 'li',
			className: () => {
				return this.options.itemClass;
			}
		});

		this.$typeAheadEl.find(this.options.list).append(item.render().el);
	}

	/**
	 * Render view
	 *
	 * @public
	 */
	render() {
		let data = {
			typeAheadContextClass: this.options.contextClass,
			typeAheadModifierClass: this.options.modifierClass
		};

		if (this.$typeAheadEl && this.$typeAheadEl.length) {
			this.$typeAheadEl.find(this.options.list).empty();
		} else {
			this.$typeAheadEl = $($.parseHTML(this.template(data)));
			this.calculateWidthAndPos();
			this.$appendTarget.append(this.$typeAheadEl);
		}

		this.collection.forEach(this.renderOne, this);

		// Maintains chainability
		return this;
	}
}

// Returns the view class
export default TypeAheadView;

scripts/type-ahead.js
/**
 * This module offers type-ahead functionality for input fields.
 *
 * @module TypeAhead
 * @version v5.1.0
 *
 * @author José Medina
 */

// Imports
import { Veams } from 'app.veams';
import VeamsComponent from 'veams/lib/common/component';
import AjaxService from 'veams/lib/services/http';
import Toggler from '../../../components/toggler/scripts/toggler';

// Variables
const $ = Veams.$;

class TypeAhead extends VeamsComponent {

	/**
	 * Constructor for our class
	 *
	 * @see module.js
	 *
	 * @param {Object} obj - Object which is passed to our class
	 * @param {Object} obj.el - element which will be saved in this.el
	 * @param {Object} obj.options - options which will be passed in as JSON object
	 */
	constructor(obj) {
		let options = {
			currentValueClass: "type-ahead__current-value",
			inputSel: '[data-js-item="type-ahead__input"]',
			minLength: 2,
			minLoadTime: 0,
			suggestionItemSel: '[data-js-item="suggestion-item"]',
			suggestionListSel: '[data-js-item="type-ahead__list"]',
			templates: {
				tplSuggestionsA11y: "SUGGESTIONS__A11Y",
				tplSuggestionsOpt: "SUGGESTIONS__OPTIONS"
			}
		};

		super(obj, options);
	}


	/**
	 * Get module information
	 */
	static get info() {
		return {
			version: '5.1.0',
			vc: true,
			mod: false // set to true if source was modified in project
		};
	}


	/**
	 * Subscribe handling
	 */
	get subscribe() {
		return {}
	}


	/**
	 * Initialize the view
	 *
	 */
	initialize() {
		this.$suggestionList = this.$el.find(this.options.suggestionListSel);
		this.$input = this.$el.find(this.options.inputSel);
		this.currentValue = this.$input.val();

		this.suggestionList = new Toggler({
			el: this.$suggestionList[0],
			namespace: 'Toggler',
			appInstance: Veams
		});

		this.suggestionList.close();

		this.service = new AjaxService(this.options.serviceOptions);

		this.service.requestDidOpen = (request, obj) => {
			if (this.options.serviceOptions.contentType) {
				request.setRequestHeader("Content-type", this.options.serviceOptions.contentType);
			}
			request.setRequestHeader("X-CSRFToken", this.options.serviceOptions["X-CSRFToken"]);
		}
	}


	/**
	 * Bind Vent events for the resize events and the load more button
	 */
	bindEvents() {
		this.$input.on(Veams.EVENTS.keyup, this.onKeyup.bind(this));
		$(window).on(Veams.EVENTS.keydown, this.onWindowKeydown.bind(this));
		this.$input.on(Veams.EVENTS.blur, this.onItemBlur.bind(this));
	}


	/**
	 * Set focus when needed
	 *
	 * @param {Object} obj - CTA event object
	 */
	setFocus(obj) {

		if (obj.isActive) {

			setTimeout(() => {
				this.$input[0].focus();
			}, 1000);
		}
	}


	/**
	 * Handle window key down event to control the module with arrow keys, enter and escape
	 *
	 * @param {Object} event - Event object
	 */
	onWindowKeydown(event) {
		if (this.el.contains(event.target)) {
			let getFocusableElements = () => {
				return Array.prototype.slice.call($("a, input", this.$el));
			};
			let focusableEls, index;

			switch (event.keyCode) {
				case 38:
					event.preventDefault();
					focusableEls = getFocusableElements();
					index = focusableEls.indexOf(document.activeElement);

					if (index > -1 && index - 1 >= 0) {
						focusableEls[index - 1].focus();
					}
					break;
				case 40:
					event.preventDefault();
					focusableEls = getFocusableElements();
					index = focusableEls.indexOf(document.activeElement);

					if (index > -1 && index + 1 < focusableEls.length) {
						focusableEls[index + 1].focus();
					}
					break;
				case 27:
					this.closeSuggestionList();
					break;
				case 13:
					this.triggerChosenSuggestion(event.target);
					this.closeSuggestionList();
					break;
			}
		}
	}


	/**
	 * Handle input key up event to control the module with arrow keys, enter and escape
	 *
	 * @param {Object} event - Event object
	 */
	onKeyup(event) {
		switch (event.keyCode) {
			case 38:
			case 40:
			case 27:
			case 13:
				event.preventDefault();
				break;
			default:
				this.currentValue = this.$input.val();
				this.render(this.generateTimeOfRequest());
		}
	}


	/**
	 * Bind filter items events as soon as render occurs
	 *
	 */
	bindItemEvents() {
		this.$suggestionList.find(this.options.suggestionItemSel)
			.on(Veams.EVENTS.touchstart + ' ' + Veams.EVENTS.click, this.onChosenSuggestion.bind(this))
			.on(Veams.EVENTS.blur, this.onItemBlur.bind(this))
			.on(Veams.EVENTS.focus, this.onItemFocus.bind(this))
			.on(Veams.EVENTS.mouseenter, this.onItemMouseEnter.bind(this))
			.on(Veams.EVENTS.mouseleave, this.onItemMouseLeave.bind(this));
	}


	/**
	 * Handle touch and click event
	 *
	 * @param {Object} event - Event object
	 * @param {object} [currentTarget] - Target to which listener was attached (used in VeamsQuery).
	 */
	onChosenSuggestion(event, currentTarget) {
		let target = currentTarget || event.currentTarget;

		event.preventDefault();

		this.triggerChosenSuggestion(target);

		this.closeSuggestionList();
	}


	/**
	 * Handle item focus event
	 *
	 * @param {Object} event - Event object
	 */
	onItemFocus(event) {
		this.$input.val($(event.target).closest(this.options.suggestionItemSel).attr('data-value'));
	}


	/**
	 * Handle item mouse enter event
	 *
	 * @param {Object} event - Event object
	 */
	onItemMouseEnter(event) {
		this.onItemFocus.apply(this, arguments);
	}


	/**
	 * Handle item mouse leave event
	 */
	onItemMouseLeave() {
		this.$input.val(this.currentValue);
	}


	/**
	 * Handle item blur event
	 *
	 * @param {Object} event - Event object
	 */
	onItemBlur(event) {
		this.$input.val(this.currentValue);

		if (!!event.relatedTarget && !this.$suggestionList[0].contains(event.relatedTarget) &&
			event.relatedTarget !== this.$input[0]) {
			this.closeSuggestionList();
		}
	}


	/**
	 * Helper function that looks for the current input value and transmit it
	 *
	 * @param {Object} target - DOM Node
	 */
	triggerChosenSuggestion(target) {
		this.currentValue = $(target).attr('data-value') || $(target).val();
		this.$input.val(this.currentValue);

		this.$el.trigger(Veams.EVENTS.typeAhead.suggestionChoosen, {
			search: this.currentValue
		});
	}


	/**
	 * Render class
	 *
	 * Perform and server request to get the suggestions to a given value
	 *
	 * @param {Number} timeOfRequest - Time of request
	 */
	render(timeOfRequest) {
		if (this.currentValue.length < this.options.minLength || !this.$el[0].contains(document.activeElement)) {
			this.closeSuggestionList();
			return false;
		}

		return new Promise((resolve, reject) => {
			this.service[this.options.serviceOptions.method]({
				data: JSON.stringify({}),
				url: this.options.serviceOptions.url + "?type-ahead=" + this.currentValue
			}).then((data) => {
				setTimeout(() => {
					if (timeOfRequest !== this.timeOfRequest) {
						return false;
					}

					this.$suggestionList.html(Veams.templater.render(this.options.templates.tplSuggestionsA11y,
						this.formatResponseDataHelper(data)));

					this.bindItemEvents();

					this.suggestionList.open();

					resolve();
				}, Math.max(this.options.minLoadTime - (Date.now() - timeOfRequest), 0));
			}, (error) => {
				reject();
				throw new Error("Typeahead render() :: could not resolve server request :: ", error);
			});
		});
	}


	/**
	 * Cache the current timestamp to figure out which was the last request
	 *
	 * If 2 delta time between request is less than minLoadTime, they will be rejected
	 * and the one with the latest timestamp will be sent
	 *
	 * @return {Number} - Time of request
	 */
	generateTimeOfRequest() {
		this.timeOfRequest = Date.now();

		return this.timeOfRequest;
	}


	/**
	 * Data formatter that prepares the data for the template to render
	 *
	 * @param {Object} data - Data to be prepared for rendering
	 * @return {Object} - Prepared data (ready for rendering)
	 */
	formatResponseDataHelper(data) {
		data.suggestions = data.suggestions.map((item) => {
			item.suggestionText = this.getSuggestionTextHelper(item.name);
			item.searchVal = this.currentValue;

			return item;
		});

		return data;
	}


	/**
	 * Data formatter that prepares the data for the template to render
	 *
	 * @param {String} name - Item name
	 * @return {String} - Modified data
	 */
	getSuggestionTextHelper(name) {
		let currentValueCaseInsensitive = new RegExp(this.currentValue, "ig");

		return name.replace(currentValueCaseInsensitive, '<em class="' + this.options.currentValueClass + '">$&</em>');
	}


	/**
	 * Close the suggestion list and empty the list
	 *
	 */
	closeSuggestionList() {
		this.suggestionList.close();
		this.$suggestionList.empty();
	}
}

export default TypeAhead;
